src/Features/Core/Portable/IncrementalCaches/SymbolTreeInfoIncrementalAnalyzerProvider.cs

```diff
@@ -221,32 +249,48 @@ private async Task UpdateSourceSymbolTreeInfoAsync(Project project, Cancellation
 
                     // Mark that we're up to date with this project.  Future calls with the same 
                     // semantic version can bail out immediately.
-                    _projectToInfo.AddOrUpdate(project.Id, projectInfo, (_1, _2) => projectInfo);
+                    _projectToInfo[project.Id] = projectInfo;
                 }
             }
 
-            private Task GetTask(Project project, Func<Task> func, CancellationToken cancellationToken)
+            [PerformanceSensitive("https://github.com/dotnet/roslyn/issues/33172", AllowGenericEnumeration = false, AllowCaptures = false)]
+            private Task UpdateReferencesAsync(Project project, CancellationToken cancellationToken)
             {
+                // Process all metadata references. If it remote workspace, do this in parallel.
+                var tasks = ArrayBuilder<Task>.GetInstance(project.MetadataReferences.Count);
                 var isRemoteWorkspace = project.Solution.Workspace.Kind == WorkspaceKind.RemoteWorkspace;
-                return isRemoteWorkspace
-                    ? Task.Run(func, cancellationToken)
-                    : func();
-            }
 
-            private Task UpdateReferencesAync(Project project, CancellationToken cancellationToken)
-            {
-                // Process all metadata references. If it remote workspace, do this in parallel.
-                var tasks = new List<Task>();
+                Debug.Assert(project.MetadataReferences is ImmutableArray<MetadataReference>, "We rely on specific collection types to ensure allocations are avoided.");
 
-                foreach (var reference in project.MetadataReferences.OfType<PortableExecutableReference>())
+                foreach (var metadataReference in project.MetadataReferences.ToImmutableArrayOrEmpty())
```

Comment:
@sharwell I think this is the only change that didn't get PR'd in your PRs. Is it worth backporting and applying it?